inode.c

```diff
@@ -533,11 +557,113 @@ static int simplefs_rmdir(struct inode *dir, struct dentry *dentry)
+    struct simplefs_inode_info *ci_dir = SIMPLEFS_INODE(dir);
+    struct simplefs_dir_block *dir_block;
+    struct buffer_head *bh;
+    int f_pos = -1, ret = 0, i = 0;
+
+    bh = sb_bread(sb, ci_dir->index_block);
+    if (!bh)
+        return -EIO;
+    dir_block = (struct simplefs_dir_block *) bh->b_data;
+
+    if (dir_block->files[SIMPLEFS_MAX_SUBFILES - 1].inode != 0) {
+        ret = -EMLINK;
+        printk(KERN_INFO "directory is full");
+        goto end;
+    }
+
+    for (i = 0; i < SIMPLEFS_MAX_SUBFILES; i++) {
+        if (dir_block->files[i].inode == 0) {
+            f_pos = i;
+            break;
+        }
+    }
+
+    dir_block->files[f_pos].inode = inode->i_ino;
+    strncpy(dir_block->files[f_pos].filename, dentry->d_name.name,
+            SIMPLEFS_FILENAME_LEN);
+    mark_buffer_dirty(bh);
+
+    inode_inc_link_count(inode);
+    d_instantiate(dentry, inode);
+end:
+    brelse(bh);
+    return ret;
+}
+
+static int simplefs_symlink(struct inode *dir,
+                            struct dentry *dentry,
+                            const char *symname)
+{
+    struct super_block *sb = dir->i_sb;
+    unsigned int l = strlen(symname) + 1;
+    struct inode *inode = simplefs_new_inode(dir, S_IFLNK | S_IRWXUGO);
+    struct simplefs_inode_info *ci = SIMPLEFS_INODE(inode);
+    struct simplefs_inode_info *ci_dir = SIMPLEFS_INODE(dir);
+    struct simplefs_dir_block *dir_block;
+    struct buffer_head *bh;
+    int f_pos = 0, i = 0;
+
+    if (l > sizeof(ci->i_data))
+        return -ENAMETOOLONG;
+
+    /* fill directory data block */
+    bh = sb_bread(sb, ci_dir->index_block);
+
+    if (!bh)
+        return -EIO;
+    dir_block = (struct simplefs_dir_block *) bh->b_data;
+
+    if (dir_block->files[SIMPLEFS_MAX_SUBFILES - 1].inode != 0) {
+        printk(KERN_INFO "directory is full\n");
+        return -EMLINK;
+    }
+
+    for (i = 0; i < SIMPLEFS_MAX_SUBFILES; i++) {
+        if (dir_block->files[i].inode == 0) {
+            f_pos = i;
+            break;
+        }
+    }
+
+    dir_block->files[f_pos].inode = inode->i_ino;
+    strncpy(dir_block->files[f_pos].filename, dentry->d_name.name,
+            SIMPLEFS_FILENAME_LEN);
+    mark_buffer_dirty(bh);
+    brelse(bh);
+
+    inode->i_link = (char *) ci->i_data;
+    memcpy(inode->i_link, symname, l);
+    inode->i_size = l - 1;
+    mark_inode_dirty(inode);
+    d_instantiate(dentry, inode);
+
+    return 0;
+}
+
+static const char *simplefs_get_link(struct dentry *dentry,
+                                     struct inode *inode,
+                                     struct delayed_call *done)
+{
+    return inode->i_link;
+}
+
 static const struct inode_operations simplefs_inode_ops = {
     .lookup = simplefs_lookup,
     .create = simplefs_create,
     .unlink = simplefs_unlink,
     .mkdir = simplefs_mkdir,
     .rmdir = simplefs_rmdir,
     .rename = simplefs_rename,
-};
+    .link = simplefs_link,
+    .symlink = simplefs_symlink};
```

Comment:
Append extra `,` close to `.symlink = simplefs_symlink`, so that `clang-format` can indent as expected.